test/Test/Database/LSMTree/Internal/Vector/Growing.hs

```diff
@@ -0,0 +1,212 @@
+        [
+            testProperty "Final vector is correct"
+                         prop_finalVectorIsCorrect
+        ]
+
+-- * Utilities
+
+-- ** Segments
+
+data Segment a = Replicate Int a deriving stock Show
+
+segmentLength :: Segment a -> Int
+segmentLength (Replicate count _) = max 0 count
+
+segmentToList :: Segment a -> [a]
+segmentToList (Replicate count val) = replicate count val
+
+appendSegment :: GrowingVector s a -> Segment a -> ST s ()
+appendSegment vec (Replicate count val) = append vec count val
+
+-- ** Vector construction
+
+{-
+    Constructs an ordinary vector by creating a growing vector, appending
+    segments to it, and finally freezing it.
+-}
+finalVector :: Int         -- Initial buffer size
+            -> [Segment a] -- Segments
+            -> Vector a    -- Final vector
+finalVector initialBufferSize segments = runST $ do
+    vec <- new initialBufferSize
+    mapM_ (appendSegment vec) segments
+    freeze vec
+
+{-
+    Supplies the final contents of a growing vector for constructing a property.
+
+    The resulting property additionally provides statistics about the buffer
+    size scaling exponents, where a buffer size scaling exponent is the binary
+    logarithm of the ratio between the buffer size after and before appending a
+    segment.
+-}
+withFinalVector
+    :: Testable prop
+    => Int                -- Initial buffer size
+    -> [Segment a]        -- Segments
+    -> (Vector a -> prop) -- Property from final vector
+    -> Property           -- Resulting property
+withFinalVector initialBufferSize segments consumer
+    = tabulate
+          "Buffer size scaling exponents"
+          (show <$> bufferSizeScalingExponents availableBufferSizes 0 segments)
+          (consumer $ finalVector initialBufferSize segments)
+    where
+
+    availableBufferSizes :: [Int]
+    availableBufferSizes = iterate (* 2) initialBufferSize
+    {-
+        We do not need to account for overflow here, because the lengths of the
+        vectors we construct are small compared to @maxBound :: Int@.
+    -}
+
+    bufferSizeScalingExponents :: [Int] -> Int -> [Segment a] -> [Int]
+    bufferSizeScalingExponents _ _ []
+        = []
+    bufferSizeScalingExponents availableSizes length (head : tail)
+        = List.length insufficient :
+          bufferSizeScalingExponents sufficient length' tail
+        where
+
+        length' :: Int
+        !length' = length + segmentLength head
+
+        insufficient, sufficient :: [Int]
+        (insufficient, sufficient) = span (< length') availableSizes
+
+-- * Properties to test
+
+prop_finalVectorIsCorrect :: Positive (Small Int)
+                          -> Segments Integer
+                          -> Property
+prop_finalVectorIsCorrect (Positive (Small initialBufferSize))
+                          (Segments segments)
+    = withFinalVector initialBufferSize segments $ \ vec ->
+      toList vec === concatMap segmentToList segments
```

Comment:
Or alternatively, compare on vectors. That is, we convert the segments to vector and concat them into a final vector, and then do the `===` check

Reply:
What would be the advantage of comparing vectors? The present solution should potentially use less memory at least, since only one vector is constructed and the lists can be processed lazily.